aws-distro-opentelemetry-node-autoinstrumentation/src/exporter/aws/metrics/aws-cloudwatch-emf-exporter.ts

```diff
@@ -0,0 +1,626 @@
+    const allAttributes: Attributes = metricRecords.length > 0 ? metricRecords[0].attributes : {};
+
+    // Process each metric record
+    for (const record of metricRecords) {
+      const metricName = record.name;
+      if (!metricName) {
+        continue;
+      }
+
+      if (record.expHistogramData) {
+        // Base2 Exponential Histogram
+        emfLog[metricName] = record.expHistogramData;
+      } else if (record.histogramData) {
+        // Regular Histogram metrics
+        emfLog[metricName] = record.histogramData;
+      } else if (record.value !== undefined) {
+        // Gauge, Sum, and other aggregations
+        emfLog[metricName] = record.value;
+      } else {
+        diag.debug(`Skipping metric ${metricName} as it does not have valid metric value`);
+        continue;
+      }
+
+      const metricData: Metric = {
+        Name: metricName,
+      };
+
+      const unit = this.getUnit(record);
+      if (unit) {
+        metricData.Unit = unit;
+      }
+      metricDefinitions.push(metricData);
+    }
+
+    const dimensionNames = this.getDimensionNames(allAttributes);
+
+    for (const [name, value] of Object.entries(allAttributes)) {
+      emfLog[name] = value?.toString() ?? 'undefined';
+    }
+
+    if (dimensionNames && metricDefinitions.length > 0) {
+      emfLog._aws.CloudWatchMetrics.push({
+        Namespace: this.namespace,
+        Dimensions: [dimensionNames],
+        Metrics: metricDefinitions,
+      });
+    }
+
+    return emfLog;
+  }
+
+  /**
+   * Group metric record by attributes and timestamp.
+   *
+   * @param record The metric record
+   * @param timestampMs The timestamp in milliseconds
+   * @returns {[string, number]} Values for the key to group metrics
+   */
+  private groupByAttributesAndTimestamp(record: MetricRecord): [string, number] {
+    // Create a key for grouping based on attributes
+    const attrsKey = this.getAttributesKey(record.attributes);
+    return [attrsKey, record.timestamp];
+  }
+
+  /**
+   * Method to handle safely pushing a MetricRecord into a Map of a Map of a list of MetricRecords
+   *
+   * @param groupedMetrics
+   * @param groupAttribute
+   * @param groupTimestamp
+   * @param record
+   */
+  private pushMetricRecordIntoGroupedMetrics(
+    groupedMetrics: Map<string, Map<number, MetricRecord[]>>,
+    groupAttribute: string,
+    groupTimestamp: number,
+    record: MetricRecord
+  ) {
+    let metricsGroupedByAttribute = groupedMetrics.get(groupAttribute);
+    if (!metricsGroupedByAttribute) {
+      metricsGroupedByAttribute = new Map<number, MetricRecord[]>();
+      groupedMetrics.set(groupAttribute, metricsGroupedByAttribute);
+    }
+
+    let metricsGroupedByAttributeAndTimestamp = metricsGroupedByAttribute.get(groupTimestamp);
+    if (!metricsGroupedByAttributeAndTimestamp) {
+      metricsGroupedByAttributeAndTimestamp = [];
+      metricsGroupedByAttribute.set(groupTimestamp, metricsGroupedByAttributeAndTimestamp);
```

Comment:
q: why not just have a flatten map with {groupAttribute+ groupTimestamp, record} instead of nested map? non-blocking